File: lib/facter/util/resolvers/aws_token.rb
# frozen_string_literal: true

module Facter
  module Util
    module Resolvers
      module AwsToken
        attr_reader :token

        @log = Facter::Log.new(self)

        class << self
          AWS_API_TOKEN_URL = 'http://169.254.169.254/latest/api/token'

          def get(lifetime = 100)
            @expiry ||= Time.now

            return @token if @token && @expiry > Time.now

            @token = nil
            @expiry = Time.now + lifetime

            headers = {
              'X-aws-ec2-metadata-token-ttl-seconds' => lifetime.to_s
            }

            @token = Facter::Util::Resolvers::Http.put_request(AWS_API_TOKEN_URL, headers)
          end

          def reset
            @expiry = nil
            @token = nil
          end
        end
      end
    end
  end
end

File: lib/src/util/aws_token.cc
#include <facter/util/aws_token.hpp>
#include <leatherman/logging/logging.hpp>

#ifdef USE_CURL
#include <leatherman/curl/client.hpp>
namespace lth_curl = leatherman::curl;

using namespace std;

namespace facter { namespace util {

    string get_token(string const& url, lth_curl::client& cli, int const& lifetime){
        lth_curl::request req(url);
        req.add_header("X-aws-ec2-metadata-token-ttl-seconds", to_string(lifetime));
        auto response = cli.put(req);

        if (response.status_code() != 200){
            LOG_DEBUG("request for {1} returned a status code of {2}.", req.url(), response.status_code());
            return "";
        }

        return response.body();
    }
}}
#endif  // USE_CURL
